### PR TITLE
feat: add formatUsdAmount, parseUsdAmount, and fromStroops to @meridian/shared

### LINKED ISSUE #252: [Feature] Add formatUsdAmount, parseUsdAmount, and fromStroops to @meridian/shared
## Problem

`packages/shared/src/utils.ts` only has `shortenAddress`, `sanitizeTxError`, and validation helpers. The issue #7 references `formatUsdAmount`, `parseUsdAmount`, and `fromStroops` as if they exist, but they do not. The testing issue is blocked until these are implemented.

## Location in codebase

- `packages/shared/src/utils.ts` (add the three functions)

## Tasks

- [ ] `fromStroops(stroops: bigint): string` — converts a stroops value to a decimal XLM string (1 XLM = 10,000,000 stroops)
- [ ] `formatUsdAmount(amount: number): string` — formats a number as a USD string (e.g. 1234.5 -> `$1,234.50`)
- [ ] `parseUsdAmount(value: string): number` — parses a USD-formatted string back to a number (reverse of `formatUsdAmount`)
- [ ] Export all three from `packages/shared/src/index.ts`

## Acceptance Criteria

- All three functions are exported from `@meridian/shared`
- `pnpm typecheck` passes with no errors
- No `any` types

## Notes

This unblocks #7 (unit tests for shared utilities).

### PR DESCRIPTION
**Pull request by Diyaaa-12**
## Summary
- Added `fromStroops(stroops: bigint): string` — converts stroops to decimal XLM string (1 XLM = 10,000,000 stroops), correctly handling whole, fractional, and negative values
- Added `formatUsdAmount(amount: number): string` — formats a number as a USD currency string (e.g. 1234.5 → "$1,234.50") using `Intl.NumberFormat`; throws `RangeError` for NaN/Infinity
- Added `parseUsdAmount(value: string): number | null` — parses a USD-formatted string back to a number; returns `null` for invalid or malformed input
- All three are exported from `@meridian/shared` via the existing `export * from "./utils"` in `index.ts`

## Changes from review feedback
- Fixed negative input bug in `fromStroops` (sign is now stripped before decimal path and re-applied)
- `parseUsdAmount` return type changed to `number | null` to handle parse failures explicitly instead of returning `NaN`
- `formatUsdAmount` now guards against non-finite inputs
- Added full test coverage for all three functions including edge cases

## Test plan
- [x] `pnpm lint && pnpm typecheck && pnpm test` pass locally
- [x] `pnpm --filter @meridian/shared typecheck` exits 0
- [x] `pnpm --filter @meridian/shared test` — 38 tests passing
- [x] No `any` types used

Closes #252

### CODE REVIEW

File: packages/shared/src/utils.ts [hunk 1]
```diff
@@ -66,4 +66,37 @@ export async function withRetry<T>(
   }
   throw lastErr;
 }
+/**
+ * Converts a stroops value to a decimal XLM string.
+ * 1 XLM = 10,000,000 stroops.
+ */
+export function fromStroops(stroops: bigint): string {
```
Comment (collinsezedike) [inline packages/shared/src/utils.ts:73]: `bigint %` inherits the sign of the dividend, so for `fromStroops(-15_500_000n)`: `remainder` is `-5_500_000n`, `remainder.toString()` is `"-5500000"` (already 8 chars), `padStart(7, "0")` is a no-op, and `.replace(/0+$/,  "")` strips trailing zeros from the numeric digits, leaving `"-55"`. The return value is `"-1.-55"` instead of `"-1.55"`. Every non-exact-unit negative input is wrong.

Fix: strip the sign before the decimal path, operate on `stroops < 0n ? -remainder : remainder`, and prepend the sign to the final string.

File: packages/shared/src/utils.ts [hunk 2]
```diff
@@ -66,4 +66,37 @@ export async function withRetry<T>(
...
+ * e.g. 1234.5 -> "$1,234.50"
+ */
+export function formatUsdAmount(amount: number): string {
+  return new Intl.NumberFormat("en-US", {
+    style: "currency",
+    currency: "USD",
+    minimumFractionDigits: 2,
+    maximumFractionDigits: 2,
+  }).format(amount);
+}
```
Comment (collinsezedike) [inline packages/shared/src/utils.ts:99]: `parseUsdAmount("")` and `parseUsdAmount("abc")` both reduce `cleaned` to `""` after the regex strip, and `parseFloat("") === NaN`. The return type is `number`, so TypeScript will not warn callers. Any downstream arithmetic on the result silently propagates `NaN`. The function should either return `number | null` (returning `null` on failure) or throw, so callers can distinguish a valid zero from a parse failure.

File: packages/shared/src/utils.ts [hunk 3]
```diff
@@ -66,4 +66,37 @@ export async function withRetry<T>(
   }
   throw lastErr;
 }
+/**
+ * Converts a stroops value to a decimal XLM string.
```
Comment (collinsezedike) [inline packages/shared/src/utils.ts]: `packages/shared/src/utils.test.ts` on main imports only `{ withRetry, withRaceTimeout, sanitizeTxError }` and the PR diff does not touch the test file. The three new exports have zero test coverage. The negative-sign bug in `fromStroops` and the NaN return in `parseUsdAmount` would both be caught immediately by a small set of unit tests. Please add cases for at minimum: `fromStroops` with a negative input, a non-exact-unit input, and zero; `parseUsdAmount` with an empty string and a multi-dot string; and `formatUsdAmount` with `NaN` and `Infinity`.

File: packages/shared/src/utils.ts (hunk 2 above)
Comment (collinsezedike) [inline packages/shared/src/utils.ts:99]: `"1.2.3".replace(/[^0-9.-]/g, "")` produces `"1.2.3"` (the regex keeps all dots), and `parseFloat("1.2.3")` returns `1.2` with no error or signal. A user pasting a corrupted value gets a wrong number back with no indication that parsing stopped early. Replacing the regex strip with a targeted match such as `value.match(/-?[0-9]+(?:\.[0-9]+)?/)?.[0]` would return `null` for malformed inputs, letting callers detect failure.

File: packages/shared/src/utils.ts [hunk 4]
```diff
@@ -66,4 +66,37 @@ export async function withRetry<T>(
...
+  const whole = stroops / 10_000_000n;
+  const remainder = stroops % 10_000_000n;
+  if (remainder === 0n) return whole.toString();
+  const decimal = remainder.toString().padStart(7, "0").replace(/0+$/, "");
+  return `${whole}.${decimal}`;
+}
+
+/**
+ * Formats a number as a USD currency string.
+ * e.g. 1234.5 -> "$1,234.50"
```
Comment (collinsezedike) [inline packages/shared/src/utils.ts:86]: `Intl.NumberFormat.format(NaN)` returns `"$NaN"` and `format(Infinity)` returns `"$∞"`. Neither throws, so a caller that passes a miscalculated number gets a visually broken but non-crashing string. Adding a guard at the top (`if (!Number.isFinite(amount)) throw new RangeError(...)` or returning a safe fallback) makes the behavior explicit.

**Comment by collinsezedike** [comment]
`VaultPanel.tsx` has a private `formatUsd` at line 13 that is functionally identical to `formatUsdAmount`. Once this PR merges, two USD formatters will coexist. Please migrate VaultPanel to import `formatUsdAmount` so there is one canonical formatter.

File: packages/shared/src/utils.ts (hunk 3 above)
Comment (collinsezedike) [inline packages/shared/src/utils.ts]: `stroopsToUnits(stroops: bigint): number` in `packages/stellar-sdk-helpers/src/defindex.ts` performs the same `/ 10_000_000n` division, returning a `number` instead of a `string`. `toStroops` (its inverse) also lives in `stellar-sdk-helpers/src/tx.ts`. Adding `fromStroops` to `@meridian/shared` splits the round-trip pair across two packages. Worth discussing whether `fromStroops` should live alongside `toStroops` in `stellar-sdk-helpers` instead.

**Comment by Diyaaa-12** [comment]
Hi @collinsezedike, I've addressed all the review comments:

- Fixed negative input bug in fromStroops
- parseUsdAmount now returns null for invalid/malformed input
- formatUsdAmount throws RangeError for NaN/Infinity
- Added test cases for all edge cases

All 38 tests passing. Please take a look when you get a chance!

File: packages/shared/src/utils.ts [hunk 5]
```diff
@@ -66,4 +66,43 @@ export async function withRetry<T>(
...
+  const abs = negative ? -stroops : stroops;
+  const whole = abs / 10_000_000n;
+  const remainder = abs % 10_000_000n;
+  const sign = negative ? "-" : "";
+  if (remainder === 0n) return `${sign}${whole}`;
+  const decimal = remainder.toString().padStart(7, "0").replace(/0+$/, "");
+  return `${sign}${whole}.${decimal}`;
+}
+
+/**
```
Comment (collinsezedike) [inline packages/shared/src/utils.ts:84]: The `Intl.NumberFormat` instance is constructed on every call to `formatUsdAmount`. The formatter options never change, so this should be a module-level constant. `Intl.NumberFormat` construction involves locale data lookup and is one of the heavier built-in allocations; instantiating it per call adds avoidable overhead in any component that formats balances on render or in a list.

**Comment by collinsezedike** [comment]
> `VaultPanel.tsx` has a private `formatUsd` at line 13 that is functionally identical to `formatUsdAmount`. Once this PR merges, two USD formatters will coexist. Please migrate VaultPanel to import `formatUsdAmount` so there is one canonical formatter.

Please migrate `VaultPanel` to import `formatUsdAmount` from `@meridian/shared` so there is one canonical formatter.

**Comment by Diyaaa-12** [comment]
Hi @collinsezedike, addressed the remaining feedback:

- Hoisted Intl.NumberFormat instance to module level as a constant to avoid per-call construction overhead
- Migrated VaultPanel.tsx to import formatUsdAmount from @meridian/shared, removing the duplicate local formatUsd

All 38 tests still passing!

**Review by collinsezedike — DISMISSED** [review]
Great turnaround on the feedback, @Diyaaa-12, coming back with all the review points addressed cleanly. 

Thank you! Love the energy.

A couple of small things on my end to tidy up before this merges.

**Comment by collinsezedike** [comment]
Great work getting this to prod, @Diyaaa-12. 

The utilities are clean, well-tested, and the iteration on review feedback was thorough. This is good to merge.

You are clear to continue with #7.
